apps/spruce/src/pages/waterfall/useFilters.ts

```diff
@@ -117,56 +60,77 @@ export const useFilters = ({
       }
     };
 
+    const activeVersions = flattenedVersions.filter(
+      (v) => v.activated && matchesRequesters(v, requesters),
+    );
+
     buildVariants.forEach((bv) => {
       const passesBVFilter =
         !buildVariantFilterRegex.length ||
         buildVariantFilterRegex.some((r) => bv.displayName.match(r));
-      if (passesBVFilter) {
-        if (
-          activeVersionIds.length !== bv.builds.length ||
-          taskFilterRegex.length
-        ) {
-          const activeBuilds: Build[] = [];
-          bv.builds.forEach((b) => {
-            if (activeVersionIds.includes(b.version)) {
-              if (taskFilterRegex.length) {
-                const activeTasks = b.tasks.filter((t) =>
-                  taskFilterRegex.some((r) => t.displayName.match(r)),
-                );
-                activeBuilds.push({
-                  ...b,
-                  tasks: activeTasks,
-                });
-              } else {
-                activeBuilds.push(b);
+
+      if (!passesBVFilter) {
+        return;
+      }
+
+      if (requesters.length || taskFilterRegex.length) {
+        const activeBuilds: Build[] = [];
+        bv.builds.forEach((b) => {
+          if (activeVersions.find(({ id }) => id === b.version)) {
+            if (taskFilterRegex.length) {
+              const activeTasks = b.tasks.filter((t) =>
+                taskFilterRegex.some((r) => t.displayName.match(r)),
+              );
+              if (activeTasks.length) {
+                activeBuilds.push({ ...b, tasks: activeTasks });
               }
+            } else {
+              activeBuilds.push(b);
             }
-          });
-          if (
-            activeBuilds.length &&
-            activeBuilds.some((b) => b.tasks.length > 0)
-          ) {
-            pushVariant({ ...bv, builds: activeBuilds });
           }
-        } else {
-          pushVariant(bv);
+        });
+        if (activeBuilds.length) {
+          pushVariant({ ...bv, builds: activeBuilds });
         }
+      } else {
+        pushVariant(bv);
       }
     });
     return bvs;
   }, [
-    activeVersionIds,
     buildVariantFilterRegex,
     buildVariants,
+    flattenedVersions,
     hasFilters,
     pins,
+    requesters,
     taskFilterRegex,
   ]);
 
+  const activeVersions = useMemo(() => {
+    const hasActiveBuild = (versionId: string) =>
+      filteredBuildVariants.some((bv) =>
+        bv.builds.some((build) => build.version === versionId),
+      );
+
+    return groupInactiveVersions(flattenedVersions, hasActiveBuild);
+  }, [filteredBuildVariants, flattenedVersions]);
```

Comment:
I was wondering what you think about moving the requester logic down here, it would be like:

```
  const activeVersions = useMemo(() => {
    const hasActiveBuild = (version: WaterfallVersionFragment) =>
      matchesRequesters(version, requesters) &&
      filteredBuildVariants.some((bv) =>
        bv.builds.some((build) => build.version === version.id),
      );
    return groupInactiveVersions(flattenedVersions, hasActiveBuild);
  }, [filteredBuildVariants, flattenedVersions, requesters]);
```

I think it would mean that you can remove all of the requester logic above

Reply:
This would complicate the render slightly by having a different number of builds vs active versions. I was able to get it to work, but it makes the useFilter test results slightly more confusing by including builds in the results that are not actually rendered. I'm fine with going with this approach if you think this tradeoff makes sense, but I was a bit undecided myself. Good callout though!